### PR TITLE
Add formatting helper for SmartUnions

### PR DESCRIPTION
**Pull request by gavin-norman-sociomantic**
A small idea that just occurred to me.

### CODE REVIEW

**Comment by gavin-norman-sociomantic** [comment]
Notes:
1. I wonder whether we should format some information about the smart union itself, rather than just about the active field. e.g. (using the union in the unittest) instead of `"23"`, we could format something like `"SmartUnion: s1: 23"`. Thoughts?
2. Needs release notes, if we decide to merge this.

**Comment by gavin-norman-sociomantic** [comment]
>I wonder whether we should format some information about the smart union itself, rather than just about the active field. e.g. (using the union in the unittest) instead of "23", we could format something like "SmartUnion: s1: 23".

Actually, I don't think we should do this. The name of the active field can already be accessed, via the `active_name` method, so if the user wants to format this information, they can do it themselves. e.g. `sformat(sink, "{}: {}", u.active_name, u)`.

**Comment by mihails-strasuns-sociomantic** [comment]
I also don't like addition of non-trivial `toString` methods as those make everything depend on formatter (and `SmartUnion` is a primitive core module on its own). See also https://github.com/sociomantic-tsunami/ocean/pull/330

**Comment by gavin-norman-sociomantic** [comment]
>I also don't like addition of non-trivial toString methods as those make everything depend on formatter (and SmartUnion is a primitive core module on its own). See also #330

I see what you mean. I don't really like the solution of having to have external functions to connect a container to the formatter :/ A third way would be to build support for containers into the formatter itself.

None of the solutions really seems ideal.

**Comment by mihails-strasuns-sociomantic** [comment]
> I don't really like the solution of having to have external functions to connect a container to the formatter

It is the only reasonable option from the module system PoV I can think of. It may feel confusing from OOP mindset, but OOP sucks anyway so no big loss. `toString` is inherently leaky abstraction as there are always many useful ways to format non-trivial data structure.

Note that fomatter will be able to print most structs without any `toString` method, you only use wrappers to specify some specific pretty format.

**Comment by gavin-norman-sociomantic** [comment]
Where would you suggest locating such a function? I see that, in #330, the formatting function (`asPrettyStr`) isn't in the same module as the container that it formats info about (`Time`). How is the user supposed to know this? It doesn't seem very user-friendly to me.

**Comment by gavin-norman-sociomantic** [comment]
Ah I see. The container is in `ocean.time.Time` and the function to format it is in `ocean.text.convert`. That does make sense.

**Comment by gavin-norman-sociomantic** [comment]
So the suggestion would be then to add `ocean.text.convert.SmartUnion`. (I don't like the package name `convert`, by the way...)

**Comment by mihails-strasuns-sociomantic** [comment]
I see two options:

1) Keep using current structure and add more formatter modules to `ocean.text.convert` (like you have described). In that case moving all new formatting code + wrappers to new `ocean.text.formatter` package could be a way to make structure cleaner.

2) Start using convention of `ocean.core.SmartUnion` and `ocean.core.SmartUnionFormatting` being two modules side by side in the same package.

Former is better overall structure, latter is easier to spot immediately.

**Comment by gavin-norman-sociomantic** [comment]
I'd go for 1. It's not so immediately obvious, but is a much cleaner structure and is easy once you're used to the idea. We can, of course, add documentation in the container packages saying "see also ocean.text.formatter".

**Comment by mihails-strasuns-sociomantic** [comment]
Ok, please add to `ocean.text.convert` for now in that case and create separate issue about `formatter` package.

**Comment by gavin-norman-sociomantic** [comment]
~~Ahh, of course, in this case (`SmartUnion`), it doesn't work. The formatting code requires access to private internals.~~ I think I can make it work.

**Comment by nemanja-boric-sociomantic** [comment]
But what about this:

> Actually, I don't think we should do this. The name of the active field can already be accessed, via the active_name method, so if the user wants to format this information, they can do it themselves. e.g. sformat(sink, "{}: {}", u.active_name, u).

?

**Comment by gavin-norman-sociomantic** [comment]
That was about printing `SmartUnion: s1: 23` vs `23`.

**Comment by gavin-norman-sociomantic** [comment]
Updated. (Still no release notes. Will add before finalising.)

**Comment by gavin-norman-sociomantic** [comment]
Updated: D2 stomping prevention fix.

**Comment by mathias-lang-sociomantic** [comment]
> A small idea that just occurred to me.

That's what I suggested when you originally implemented it in swarm BTW :D
I think this approach removes most of the convenience of the formatter though, as you have to explicitly say you want a "pretty" string. It also completely kills generic code.

**Comment by gavin-norman-sociomantic** [comment]
>I think this approach removes most of the convenience of the formatter though, as you have to explicitly say you want a "pretty" string. It also completely kills generic code.

I feel the same way.

**Comment by gavin-norman-sociomantic** [comment]
>That's what I suggested when you originally implemented it in swarm BTW :D

I don't remember why I didn't follow your suggestion then...

**Comment by gavin-norman-sociomantic** [comment]
@mathias-lang-sociomantic the "third way" would be to add support for smart union formatting _inside_ `Formatter`. Your thoughts on that?

**Comment by mathias-lang-sociomantic** [comment]
@gavin-norman-sociomantic : I would be very opposed to it. It sounds very backward to modify a formatter to support a type, because there's many more type to support than formatting method.

The `toString` approach was specifically designed to not depend directly on the Formatter. The reason one has to import `Formatter` is because we need to convert a type to its string representation (like int => string). In the case of `SmartUnion`, you want a generic method, so there's no way around the Formatter.

However, I added my notes on how we can address the perceived problem of the Formatter being a dependency [in the related issue](https://github.com/sociomantic-tsunami/ocean/issues/340#issuecomment-341707137).

File: src/ocean/text/convert/SmartUnion.d [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+        callWithActive function in ocean.core.SmartUnion. It is not intended to
+        be called by users.
+
+        Params:
+            T = type of union member to format
+            union_member = union member to format
+
+    ***************************************************************************/
+
+    static public void formatUnionMember ( T ) ( T union_member )
```
Comment (nemanja-boric-sociomantic) [inline src/ocean/text/convert/SmartUnion.d]: Maybe add comment `// package` here, so we can handle these cases one day.

**Comment by gavin-norman-sociomantic** [comment]
Re: #340, thoughts on changing this PR to add the new module in `ocean.text.formatter` already?

**Comment by gavin-norman-sociomantic** [comment]
Updated:
* Rebased.
* Added `package`.
* Fixed TODO and incorrect doc.
* Added release notes.

File: relnotes/smartunionformatter.feature.md [hunk 2]
```diff
@@ -0,0 +1,8 @@
+### New `SmartUnion` formatting function
+
+`ocean.text.convert.SmartUnion`
+
+The helper function `formatActive` accepts a smart-union and returns a struct
```
Comment (mihails-strasuns-sociomantic) [inline relnotes/smartunionformatter.feature.md]: Maybe `asActive` to match https://github.com/sociomantic-tsunami/ocean/blob/v3.x.x/src/ocean/text/convert/DateTime_tango.d#L968 ?

**Comment by mihails-strasuns-sociomantic** [comment]
> thoughts on changing this PR to add the new module in ocean.text.formatter already?

Depends on how fast would you like to have this functionality. Considering how long this PR was rotting, it is probably not hat urgent and we can proceed with `formatter` already.

File: relnotes/smartunionformatter.feature.md (hunk 2 above)
Comment (gavin-norman-sociomantic) [reply relnotes/smartunionformatter.feature.md]: Hmm... that's not a name that makes any sense to the casual reader.

Comment (mihails-strasuns-sociomantic) [reply relnotes/smartunionformatter.feature.md]: Well, having `format` as part of name is definitely redundant and confusing as you can only ever use this wrapper with `(s)format` calls. It affects representation, not actually formats stuff.

I am not sure what would be the best naming pattern though.

Comment (gavin-norman-sociomantic) [reply relnotes/smartunionformatter.feature.md]: I'd suggest simply `activeField`, so the usage would look like:
```D
sformat(buf, "union active member: {}", activeField(su));
```

Comment (mihails-strasuns-sociomantic) [reply relnotes/smartunionformatter.feature.md]: How would you call the time one? I'd prefer to have consistent naming style for all formatter wrappers if possible.

Comment (gavin-norman-sociomantic) [reply relnotes/smartunionformatter.feature.md]: Maybe:
```D
sformat(buf, "the time is: {}", prettyString(time));
```

Comment (mihails-strasuns-sociomantic) [reply relnotes/smartunionformatter.feature.md]: Hm, how is that any clearer than `asActiveField` and `asPrettyString` though? From your naming one may expect that field/string gets returned eagerly and passed to formatter after that. At least that is how I read it.

Comment (gavin-norman-sociomantic) [reply relnotes/smartunionformatter.feature.md]: I don't really feel the `as` part adds much meaning, but I'm fine with calling it `asActiveField`, for the sake of consistency.

**Comment by gavin-norman-sociomantic** [comment]
Updated:
* Changed function name to `asActiveField`.
* Moved to `ocean.text.formatter`.

File: relnotes/smartunionformatter.feature.md (hunk 2 above)
Comment (mihails-strasuns-sociomantic) [reply relnotes/smartunionformatter.feature.md]: I find it useful to clearly denote "eager vs lazy" in either module names or function names, similar to how there is `split` and `splitter` in Phobos (ignoring stupid name choice itself :P)

File: src/ocean/text/formatter/SmartUnion.d [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+
+    mixin TypeofThis;
+
+    /// Smart union to format.
+    private SU smart_union;
+
+    /// Flag to toggle formatting of the name of the active union member. If
+    /// true, format output will look like "<activename>: value". If false,
+    /// format output will look like "value".
+    private bool include_name;
```
Comment (mihails-strasuns-sociomantic) [inline src/ocean/text/formatter/SmartUnion.d:87]: Is there really any reason to mark field private here, especially configuration ones?

Comment (gavin-norman-sociomantic) [reply src/ocean/text/formatter/SmartUnion.d:87]: They're not intended to be accessible to the caller of `asActiveField`. I agree that modifying this config field in particular would be harmless, but it's not the intended usage of this module.

**Review by mihails-strasuns-sociomantic — APPROVED** [review]
LGTM! Would be nice to check this doesn't contradict whatever @mathias-lang-sociomantic had in mind too much.

File: src/ocean/core/SmartUnion.d [hunk 4]
```diff
@@ -3,6 +3,10 @@
     Template for a union that knows its active field and ensures the active
     field is read.
 
+    See_Also:
+        ocean.text.convert.SmartUnion -- for helper functions to format a
```
Comment (mathias-lang-sociomantic) [inline src/ocean/core/SmartUnion.d]: `s/convert/formatter`

File: src/ocean/text/formatter/SmartUnion.d [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+    {
+        int i;
+        cstring s;
+    }
+
+    SmartUnion!(U) su;
+
+    mstring buf;
+    // sformat from ocean.text.convert.Formatter
+    sformat(buf, "union active member: {}", asActiveField(su));
```
Comment (mathias-lang-sociomantic) [inline src/ocean/text/formatter/SmartUnion.d]: The output is not tested ?
Also might as well use `format` and remove the buffer.

File: src/ocean/text/formatter/SmartUnion.d [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+        cstring s;
+    }
+
+    SmartUnion!(U) su;
+
+    // Inactive union.
+    {
+        mstring buf;
+        sformat(buf, "{}", asActiveField(su));
+        test!("==")(buf, "");
```
Comment (mathias-lang-sociomantic) [inline src/ocean/text/formatter/SmartUnion.d]: No need to use a `buf` if you don't reuse it, just call `format` instead.

File: src/ocean/text/formatter/SmartUnion.d [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+    }
+
+    // s active, plus name formatting.
+    {
+        mstring buf;
+        sformat(buf, "{}", asActiveField(su, true));
+        test!("==")(buf, "<s>: hello");
+    }
+}
+
```
Comment (mathias-lang-sociomantic) [inline src/ocean/text/formatter/SmartUnion.d]: Trailing line

File: src/ocean/text/formatter/SmartUnion.d (hunk 6 above)
Comment (mathias-lang-sociomantic) [reply src/ocean/text/formatter/SmartUnion.d]: Also all could be up inline:
```
test!("==")(format("{}", asActiveField(su)), "");
```

This way no need for blocks either.

**Comment by mathias-lang-sociomantic** [comment]
Styling fixes request, but the idea LGTM

File: src/ocean/text/formatter/SmartUnion.d (hunk 5 above)
Comment (gavin-norman-sociomantic) [reply src/ocean/text/formatter/SmartUnion.d]: It's a usage example, not a test.

**Comment by gavin-norman-sociomantic** [comment]
Updated with tweaks suggested by @mathias-lang-sociomantic.

**Review by mihails-strasuns-sociomantic — APPROVED** [review]

**Comment by gavin-norman-sociomantic** [comment]
Hmm no status from Travis...? I'll rebase and repush.

**Pull request merged**